### PR TITLE
feat: keep special props for name field for customized widget

### PR DESCRIPTION
**Pull request by liufuyang**
As we have some nice special fields added for `name` fields and we treat it will use the `string` widget, it would be nice to preserve a bunch of the customized fields for `name` field even if a customized widget is used, as the customized widget (for example an uuid widget) will be very similar like a string widget and the previously added attributes such as updated `label`, `pattern` and `required.`

So it would be nice, for example if some `uuid` widget is used on name, it can still generate output as:
```
      - name: "name"
        label: "RESOURCE NAME"
        comment: "The resource name of the kitchen sink."
        required: true
        hint: "The resource name of the kitchen sink."
        pattern:
          - "^kitchenSinks/[a-z0-9][a-z0-9-]{0,61}[a-z0-9]$"
          - "Must match ^kitchenSinks/[a-z0-9][a-z0-9-]{0,61}[a-z0-9]$"
        widget: "uuid"
        outer:
          inner: 42
```

Otherwise current version generate the name field without the nice addons:

```
      - name: "name"
        label: "NAME"
        comment: "The resource name of the kitchen sink."
        required: false
        hint: "The resource name of the kitchen sink."
        widget: "uuid"
        outer:
          inner: 42
```

This will make setting up a customized widget for the name field also easier in our private repo.

### CODE REVIEW

File: cmd/protoc-gen-netlify-cms/main.go [hunk 1]
```diff
@@ -375,8 +375,9 @@ func inferField(
 		}
 		proto.Merge(field, fieldAnnotation)
 	}
-	if field.Widget.WidgetType != nil {
-		return field, true // if a widget is specified, no further inference
+	if field.Widget.WidgetType != nil &&
+		!(protoField.Desc.Kind() == protoreflect.StringKind && !protoField.Desc.IsList()) {
```
Comment (odsod) [inline cmd/protoc-gen-netlify-cms/main.go:379]: The original idea was that either you specify a custom widget and then early-exit here, or do the inference below - with this change this assumption doesn't hold anymore.

File: cmd/protoc-gen-netlify-cms/main.go [hunk 2]
```diff
@@ -412,8 +413,10 @@ func inferField(
 		}
 		return field, true
 	case protoField.Desc.Kind() == protoreflect.StringKind && !protoField.Desc.IsList():
-		field.Widget.WidgetType = &cmsv1.Widget_StringWidget{
-			StringWidget: &cmsv1.StringWidget{},
+		if field.Widget.WidgetType == nil {
```
Comment (odsod) [inline cmd/protoc-gen-netlify-cms/main.go:416]: this looks like a side-effect of breaking the invariant of not proceeding on this code path if a widget has been annotated - if you want special casing for name fields, is there a way to do it without breaking this invariant?

File: cmd/protoc-gen-netlify-cms/main.go (hunk 1 above)
Comment (liufuyang) [reply cmd/protoc-gen-netlify-cms/main.go:379]: Yes, but only for the name field.

The annoying part is for `name` we change the label from `Name` to `Resource Name`, and required set to `true`,  and these fields has nothing to do with what widget it is.

File: cmd/protoc-gen-netlify-cms/main.go [hunk 3]
```diff
@@ -375,8 +375,9 @@ func inferField(
 		}
 		proto.Merge(field, fieldAnnotation)
 	}
-	if field.Widget.WidgetType != nil {
-		return field, true // if a widget is specified, no further inference
+	if field.Widget.WidgetType != nil &&
+		!(protoField.Desc.Kind() == protoreflect.StringKind && !protoField.Desc.IsList()) {
+		return field, true // if a widget is specified, and it is not name field, no further inference
```
[the remark this answers is not in the document]
Comment (liufuyang) [reply cmd/protoc-gen-netlify-cms/main.go:380]: Aha, I missed that, need a name check there.

File: cmd/protoc-gen-netlify-cms/main.go (hunk 2 above)
Comment (liufuyang) [reply cmd/protoc-gen-netlify-cms/main.go:415]: the code below only does special configuration for `name` field, so for normal string case it is fine?

or if you want, we can perhaps add an early return here if the field is not name and it has some widgetType already?

Comment (liufuyang) [reply cmd/protoc-gen-netlify-cms/main.go]: made an update with another commit, this looks simpler and we also keep the old flow if the field is not a `name`

**Review by odsod — APPROVED** [review]
Good stuff :star:

**Pull request merged**